### PR TITLE
LGTM fixes

### PR DESCRIPTION
**Pull request by mrc0mmand**
A few fixes for errors/warnings reported by LGTM.

I'm not sure if there's any elegant solution to https://lgtm.com/projects/g/systemd/systemd/snapshot/91378e6b80ce86ec2c9be81508c38c838eb11565/files/src/shared/firewall-util.c?sort=name&dir=ASC&mode=heatmap#x62def701dfc13e50:1 (if we even want to *fix* it) or if we should just suppress it.

### CODE REVIEW

File: src/udev/scsi_id/scsi_id.c [hunk 1]
```diff
@@ -153,7 +154,7 @@ static int get_file_options(const char *vendor, const char *model,
         char *str1;
         char *vendor_in, *model_in, *options_in; /* read in from file */
         int lineno;
-        int c;
+        int c, r = 0;
```
Comment (poettering) [inline src/udev/scsi_id/scsi_id.c]: no need to assign zero here?

File: src/udev/scsi_id/scsi_id.c [hunk 2]
```diff
@@ -180,9 +181,14 @@ static int get_file_options(const char *vendor, const char *model,
         for (;;) {
                 vendor_in = model_in = options_in = NULL;
 
-                buf = fgets(buffer, MAX_BUFFER_LEN, f);
-                if (buf == NULL)
+                r = read_line(f, MAX_BUFFER_LEN, &buf);
+                if (r < 0) {
```
Comment (poettering) [inline src/udev/scsi_id/scsi_id.c]: hmm, the existing codebase here is a bit messy, but we probably should log here on error... hence maybe:

```c
retval = log_error_errno(r, …);
```

File: src/udev/scsi_id/scsi_id.c [hunk 3]
```diff
@@ -180,9 +181,14 @@ static int get_file_options(const char *vendor, const char *model,
         for (;;) {
                 vendor_in = model_in = options_in = NULL;
 
-                buf = fgets(buffer, MAX_BUFFER_LEN, f);
-                if (buf == NULL)
+                r = read_line(f, MAX_BUFFER_LEN, &buf);
```
Comment (poettering) [inline src/udev/scsi_id/scsi_id.c]: This change means we now read everything into a buffer 'buf'. That buffer is never freed (it needs to be freed on every line). Moreover the buffer allocated as `buffer` isn't useful anymore and should be removed

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging fec8efd85bf29bf05f15135211ada3d67e8e57f1 into 60a3e982f44778142b5adce96f2caaa0a11b6d54 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-18771bfeceb3b95e9d72ec0a1a8229e380ef75b2)

**fixed alerts:**

* 1 for Use of potentially dangerous function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
@poettering thanks a lot for the feedback (~~I just realized how rusty my C actually is~~). I've rewritten the problematic sections and gave it few tries with valgrind to catch memory leaks. The whole `scsi*` stuff needs some slight refactoring imo, and I'm not actually sure if I haven't made it worse.

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging 50a8e231b1a832e3072b240110104b538c489c68 into c49357e7477564b98d977e35393f471d2088dc19 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-2629ffeaab06b71cf5f7171acfc58a7af911c7d5)

**fixed alerts:**

* 1 for Use of potentially dangerous function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging 13b2656799d6e0b194043d43b7f37664a23210ac into c49357e7477564b98d977e35393f471d2088dc19 - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-878d9bb86fea02ec8b088a6dae2e08f11b003d16)

**fixed alerts:**

* 1 for Use of potentially dangerous function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by mrc0mmand** [comment]
Just a note why the *constant comparison* alert is still present:

> You won’t see the alert disappear in the PR itself. The PR doesn’t take into account alerts or files suppressions on purpose, to avoid a malicious contribution such as some_security_backdoor(); //lgtm not being reported.
> But when the PR is merged, the alert won’t appear anymore

Quoted from https://discuss.lgtm.com/t/suppressed-alert-still-appearing/1693/2

**Comment by evverx** [comment]
This pull request **fixes 1 alert** when merging e5a50d83d3ba94955c60bb0c3fe5220af4738a21 into adca059d55fe0a126dbdd62911b0705ddf8e9b8a - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-2c6ecd2eacc67c4c526b5cc0c176b934874b1bfe)

**fixed alerts:**

* 1 for Use of potentially dangerous function

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by poettering** [comment]
I must admit, I don#t fully grok the scsi_id fixes... That code is a bit over my head, and all the strsep() use makes me quite unhappy already. I mean, that's why there is #7594: we'd really like to get rid of scsi_id eventually, and the code bitrotted... It's also the reason the code still uses fgets() as only location right now...

Any chance you can split out the other fix from this PR? tht way we can merge that quickly.

**Comment by mrc0mmand** [comment]
Honestly, the whole scsi_id code would benefit from a significant refactor. I totally missed the #7594 PR - in such case the scsi_id fixes are indeed pointless. Instead of creating a new PR, I'll drop the scsi_id *fixes* from this one a possibly focus on other issues found by LGTM in other parts (and possibly *alive* parts) of the codebase.

**Pull request merged**